feat(DPE-6715) : LDAP Manager

Gu1nness opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- markdownlint-disable MD041 -->

## 🏷 Type of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update
- [ ] Tooling and CI changes
- [ ] Dependencies upgrade or change

## 📝 Description
<!--
Describe your changes in detail including:
  - the purpose and scope of this PR (what)
  - the impacted components (where, often match conventional commit scope)
  - explanation/algorithm if require (how)
-->
This PR introduces the manager for LDAP integration.
It also introduces a new peer relation for the LDAP data. This gives a better separation of concerns of the databags.
This PR does not listen to any events, it just introduces the managers.

## 📑 Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- Either the associated JIRA reference or a paragraph. -->

This is part of the LDAP integration project.
There are two other PRs opened, #22 that adds config options for this LDAP integration and #23 that updates the configuration handling to write in config files instead of passing arguments as CLI parameters.

## 🧪 How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
This has been tested using unit tests.

## ✅ Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](../blob/6/edge/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Gu1nness committed:
wip: DPE-6715

Gu1nness committed:
fix: progess

MiaAltieri commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -150,7 +150,8 @@ def on_relation_broken(self, relation: Relation) -> None:
             logger.info("Relation broken event occurring due to scale down.")
             return
-        self.remove_ca_cert_from_trust_store()
+        self.dependent.remove_ca_cert_from_trust_store(TrustStoreFiles.PBM.value)
```
nice

MiaAltieri commented on single_kernel_mongo/core/secrets.py:
```diff
@@ -21,14 +21,16 @@
     Application Logic. To be kept separate from classes below, which are simply to provide a
     (smart) abstraction layer above Juju Secrets.
     """
     members = [app_name, scope.value]
+    if relation:
+        members = [relation, app_name, scope.value]
```
nit
```suggestion
    members = [app_name, scope.value]
    if relation:
        members.append(relation)
```

reason: easier to see the diff

Gu1nness replied:
This is not the same behavior as it would put the relation at the end, which differs from what is done everywhere else: `<relation>.<app>.<scope>`.
I agree that this is not the best diff but I'm trying to stick with the standards :)

MiaAltieri replied:
ah - thanks for the explaination. No worries then!

MiaAltieri commented on single_kernel_mongo/exceptions.py:
```diff
@@ -243,3 +243,11 @@ class ActionFailedError(Exception):
+
+
+class LDAPSNotEnabledError(Exception):
+    """Raised when the charm hasn't received ldaps urls.."""
```
```suggestion
    """Raised when the charm hasn't received ldaps urls."""
```

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,162 @@
+        if not self.state.db_initialised:
+            raise DeferrableFailedHookChecksError("DB is not initialised")
+        if self.state.is_role(MongoDBRoles.SHARD):
+            self.charm.status_manager.to_blocked("Cannot integrate LDAP with shard.")
```
nit: mind doing `set_and_share_status`, makes it easier in my status work when I create constants :)

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,162 @@
+        """Runs when LDAP is ready."""
+        self.assert_pass_hook_checks()
+
+        self.charm.status_manager.to_maintenance("Configuring LDAP")
```
same comment about statuses as above

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,162 @@
+        ldap_data = self.ldap_requirer.consume_ldap_relation_data(relation=relation)
+        if not ldap_data:
+            logger.info("Waiting for LDAP data.")
+            raise WaitingForLdapDataError("Waiting for LDAP data.")
+
+        if not ldap_data.ldaps_urls:
+            logger.info("You must configure Glauth with LDAPS support")
+            raise LDAPSNotEnabledError("LDAPS not enabled.")
```
should a status be set here in either of these cases?

Gu1nness replied:
This will be set on the event handler I think.
However I'm open to set it here if that makes your work later easier.

MiaAltieri replied:
setting it here will make my work later easier!

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,162 @@
+
+        ldap_data = self.ldap_requirer.consume_ldap_relation_data(relation=relation)
+        if not ldap_data:
+            logger.info("Waiting for LDAP data.")
```
should we defer?

Gu1nness replied:
Defer will be done in the event handler as usual.
The manager has no information of events (and I want to keep it that way) so the event handler will catch this error and defer.
If you have a look at the exception definition, you can already see that it does inherits from `DeferrableError` :)

MiaAltieri replied:
Ah - yes! thank you Neha. I appreciate your patience!

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,162 @@
+            raise WaitingForLdapDataError("Waiting for LDAP data.")
+
+        if not ldap_data.ldaps_urls:
+            logger.info("You must configure Glauth with LDAPS support")
```
```suggestion
            logger.error("You must configure Glauth with LDAPS support")
```

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,162 @@
+
+
+class LDAPManager(Object, StatusProvider):
+    """Manage the relation between glauth-k8s and mongos router."""
```
```suggestion
    """Manage the relation between glauth-k8s and replica set, config-server, or mongos-router."""
```

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,162 @@
+        self.relation_name = relation_name
+        self.cert_relation_name = cert_relation_name
+        self.ldap_requirer = LdapRequirer(self.charm, self.relation_name)
+        self.certificate_transfer = CertificateTransferRequires(self.charm, self.cert_relation_name)
```
I believe you are missing the event file to listen to:
```
        self.framework.observe(
            self.ldap_requirer.on.ldap_ready,
            self._on_ldap_ready,
        )
        self.framework.observe(
            self.ldap_requirer.on.ldap_unavailable,
            self._on_ldap_unavailable,
        )

```

Gu1nness replied:
Cf PR description:
>  This PR does not listen to any events, it just introduces the managers.

MiaAltieri replied:
oopsies duh 🤦‍♀️ - thanks for your patience!

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,162 @@
+
+        self.dependent.restart_charm_services()
+
+    def on_certificate_available(self, certificate: str, ca: str, chain: list[str]):
```
where are you listening to this event?

Gu1nness replied:
Cf PR description:
>  This PR does not listen to any events, it just introduces the managers.

MiaAltieri replied:
oopsies duh 🤦‍♀️ - thanks for your patience!

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,162 @@
+        self.state.ldap.set_certificates(None, None, None)
+        self.dependent.remove_ca_cert_from_trust_store(TrustStoreFiles.LDAP)
+
+        self.dependent.restart_charm_services()
```
If we are in one of the cases of not having the cert and having the ldap interface (and vice versa) is the charm in the broken state? i.e. can't proceed with other events? Or is it only blocking clients?

Gu1nness replied:
It will not be in a broken state, it can process other events.
However identity users will not be able to log in as ldap won't be configured on the services.

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,162 @@
+
+    def on_certificate_available(self, certificate: str, ca: str, chain: list[str]):
+        """Runs when we receive the LDAP certificates."""
+        self.assert_pass_hook_checks()
```
don't shards need these certs? Since clients eventually communicate with shards?

Gu1nness replied:
This will be done in another PR if needed.

Keep in mind that this is for identity users, not client applications.
It is not recommended to log in directly on a shard, but use mongos instead.
The decision was taken not to integrate with shards, and sharing the necessary information will be done if necessary by the config-server.

As I said, this PR introduces the manager, it does not even update the configuration of the services.
I'm trying to keep small PR sizes and minimal changes to reduce the risk of breaking things.
On top of that, since this PR does not listen to events, it is for now completely dead code which make continuous development easier, not breaking the charms.

MiaAltieri replied:
Thanks for the explanation! Gotcha!

MiaAltieri reviewed (commented):
Great work! Is it possible at this stage to test manually within the charm doing a few sanity check ops? If so that would be great! If not NW

Gu1nness commented:
> Great work! Is it possible at this stage to test manually within the charm doing a few sanity check ops? If so that would be great! If not NW

The best we can do for now is unit testing as I did + ensuring that we don't break anything with the remote integration workflow. However remote integration will still need to be updated because we need to update the metadata files to add the new relations.

Gu1nness committed:
fix: post review

Mehdi-Bendriss commented on single_kernel_mongo/config/literals.py:
```diff
@@ -122,4 +122,11 @@ class VmUser(WorkloadUser[int]):
 }
 
-TRUST_STORE_CERTIFICATE_PATH = Path("/usr/local/share/ca-certificates/pbm.crt")
+TRUST_STORE_PATH = Path("/usr/local/share/ca-certificates")
```
Can we define a custom `${SNAP_DATA}` relative localized trust store that is not system-wide? (with restricted permissions)

Gu1nness replied:
I'll investigate.
I have never done such things so I'll read a bit about that and see if that is doable or not.

Gu1nness replied:
Done in #29

Mehdi-Bendriss commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -377,7 +378,10 @@ def set_config_options(self, credentials: dict[str, str]) -> None:
         # Add certificate to trust store
-        self.save_ca_cert_to_trust_store(credentials)
+        cert_chain_list = credentials.get("tls-ca-chain", None)
+        if cert_chain_list:
```
```suggestion
        if cert_chain_list := credentials.get("tls-ca-chain", None):
```

Mehdi-Bendriss commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,175 @@
+
+        self.charm.status_manager.set_and_share_status(MaintenanceStatus("Configuring LDAP"))
+
+        ldap_data = self.ldap_requirer.consume_ldap_relation_data(relation=relation)
+        if not ldap_data:
```
nit: walrus operator

Mehdi-Bendriss commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,175 @@
+
+        self.restart_when_ready()
+
+    def restart_when_ready(self):
```
```suggestion
    def restart_when_ready(self) -> None:
```

Mehdi-Bendriss commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,175 @@
+
+        self.dependent.restart_charm_services()
+
+    def on_certificate_available(self, certificate: str, ca: str, chain: list[str]):
```
```suggestion
    def on_certificate_available(self, certificate: str, ca: str, chain: list[str]) -> None:
```

Mehdi-Bendriss commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,175 @@
+
+        self.restart_when_ready()
+
+    def on_certificate_removed(self):
```
```suggestion
    def on_certificate_removed(self) -> None:
```

Mehdi-Bendriss commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,175 @@
+
+        if self.state.ldap.is_ready():
+            logger.info("Restarting mongodb server for LDAP integration")
+            self.dependent.restart_charm_services()
```
I like that this handled by the `dependent`, for when we introduce rolling restarts

Mehdi-Bendriss commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,175 @@
+        if self.state.ldap.is_ready():
+            logger.info("Restarting mongodb server for LDAP integration")
+            self.dependent.restart_charm_services()
+            self.charm.status_manager.set_and_share_status(ActiveStatus())
```
Setting an Active Status from the ldap module seems too radical for me, can we be more targeting?

Gu1nness replied:
It's a bit tricky, because we are just restarting the mongod service so it might not be available yet.
I can add a more generic status but we know that it tends to be capricious.
TBH, I used an ActiveStatus mainly not to make any decisions here as statuses would be reworked right after by advanced statuses.

Mehdi-Bendriss commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,175 @@
+        if self.state.is_role(MongoDBRoles.SHARD):
+            return BlockedStatus("Cannot integrate LDAP with shard.")
+
+        if self.state.ldap_relation is not None and self.state.ldap_cert_relation is None:
```
The first part of the condition is implied
```suggestion
        if self.state.ldap_cert_relation is None:
```

Mehdi-Bendriss commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,175 @@
+                f"`juju integrate {self.state.ldap_relation.app.name}:send-ca-cert {self.charm.app.name}:{ExternalRequirerRelations.LDAP_CERT}`"
+            )
+            return BlockedStatus("TLS is mandatory for LDAP transport.")
+        if self.state.ldap_relation is None and self.state.ldap_cert_relation is not None:
```
Similar comment:
```suggestion
        if self.state.ldap_relation is None:
```

Mehdi-Bendriss commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,175 @@
+                return BlockedStatus("Waiting for Glauth certificates.")
+            case False, True:
+                logger.info("Waiting for LDAP data.")
+                return BlockedStatus("Missing LDAP data from Glauth.")
```
These should be WaitingStatus(es) I think - as it is just a matter of time until those are resolved, and will not require user intervention

Mehdi-Bendriss commented on single_kernel_mongo/state/ldap_state.py:
```diff
@@ -0,0 +1,196 @@
+        if not value:
+            self.secrets.remove(Scope.UNIT, LdapStateKeys.CHAIN.value)
+            return
+        self.secrets.set(LdapStateKeys.CHAIN.value, value, Scope.UNIT)
```
In another PR - We'll need to readjust the secrets API, some methods have the Scope as a first arg, where others have it as the last arg

Gu1nness replied:
Definitely, I'm always having the exact same remark in the back of my head.

Mehdi-Bendriss commented on single_kernel_mongo/core/operator.py:
```diff
@@ -248,3 +253,27 @@ def set_permissions(self) -> None:
+        # Remove the file
+        self.workload.delete(TRUST_STORE_PATH / file)
+        # Update CA certificates to remove the certificate from the trust store
+        self.workload.exec("update-ca-certificates")
```
related to the above comment on localized trust store

Mehdi-Bendriss reviewed (commented):
Thanks Neha! Great work! I have some comments to clarify

Mehdi-Bendriss commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -0,0 +1,175 @@
+        """Returns True if the integration to ldap is valid."""
+        return (self.state.ldap_relation is not None) and not self.state.is_role(MongoDBRoles.SHARD)
+
+    def on_ldap_ready(self, relation: Relation) -> None:
```
Could you rename all of these methods to a more "manager" friendly name, that indicates the actual business logic it is handling. The `on_*` naming should be reserved to the events

Gu1nness replied:
Sure !

Gu1nness committed:
fix: post-review

Gu1nness committed:
Merge branch '6/edge' into feat/DPE-6715-ldap-manager

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -113,9 +128,7 @@ def on_certificate_available(self, certificate: str, ca: str, chain: list[str]):
-        if self.state.ldap.is_ready():
-            logger.info("Restarting mongodb server for LDAP integration")
-            self.dependent.restart_charm_services()
+        self.restart_when_ready()
```
Are there any dangers of not deferring when not ready?

Gu1nness replied:
We specifically don't want to defer here:
1) we are waiting on two relations
2) restart when ready first gets the component status:
 * if there is no status, there is no integration so we don't restart
 * if there is an active status, then everything is integrated and we have checked that the connection works
 * Else, there is something blocking and we go in blocked state because this will not auto resolve.

MiaAltieri commented on single_kernel_mongo/managers/ldap.py:
```diff
@@ -79,25 +82,37 @@ def on_ldap_ready(self, relation: Relation) -> None:
         if self.charm.unit.is_leader():
             self.state.ldap.set_from(ldap_data)
 
+        self.restart_when_ready()
```
Are there any dangers of not deferring when not ready?  (I assume no since the relation data would update and trigger the event calling this function)

Gu1nness replied:
We specifically don't want to defer here:
1) we are waiting on two relations
2) restart when ready first gets the component status:
 * if there is no status, there is no integration so we don't restart
 * if there is an active status, then everything is integrated and we have checked that the connection works
 * Else, there is something blocking and we go in blocked state because this will not auto resolve.

MiaAltieri commented:
Looks great! Small comments only <3

Gu1nness committed:
fix: rename methods

Mehdi-Bendriss reviewed (dismissed):
Thanks Neha!

Gu1nness committed:
feat(DPE-6716): LDAP Events (#29)

## 🏷 Type of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update
- [ ] Tooling and CI changes
- [x] Dependencies upgrade or change

## 📝 Description
LDAP events and refactor a bit the management of restarts
This PR introduces the `LDAPEventHandler` which handles the events of the three following relations:
 * ldap
 * ldap-certificate-transfer
 * ldap-peers

It also introduces a custom ops event that is used to provide a single source of truth for restarts in the LDAP integration.
There is a bit of rework of the restarts, and an improvement of the get_status that now also checks for the ldap connection by doing a bind operation.

## 📑 Motivation and Context
This is the implementation of [DPE-6716](https://warthogs.atlassian.net/browse/DPE-6716) for the ldap integration.
It adds the support for the events.
The next steps will be to update the config so that the services are configured to integrate with LDAP.

## 🧪 How Has This Been Tested?
This has been manually tested for now.
I'm currently adding unit tests.

## ✅ Checklist
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](../blob/6/edge/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.


[DPE-6716]: https://warthogs.atlassian.net/browse/DPE-6716?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Gu1nness committed:
feat: LDAP configuration (#32)

This PR updates the configuration managers for mongodb and mongos to conditionally add the ldap configuration for the service.
It also adds some testing for this configuration and some full cycle unit testing.
Finally, it updates the restart mechanism to restart only if needed.


* feat(DPE-6792): Share config with mongos (#33)

Instead of requiring the operator to config the two services, the config-server shares the ldapUserToDNMapping with mongos through the cluster relation.
It also shares a hash of the ca chain + ldap addresses to ensure that both services are connected with the same ldap server.

---------

Co-authored-by: Mia Altieri <32723809+MiaAltieri@users.noreply.github.com>

Gu1nness committed:
fix: ellipsis

Gu1nness committed:
fix: minor fixes

Gu1nness committed:
fix: cleaner restart

Gu1nness committed:
fix: clean databag for leader only

Gu1nness committed:
fix: small change

Gu1nness committed:
fix: move ldap config to ldap peer relation (allows restarts)

Gu1nness committed:
fix: restart services as soon as we lost the relation

Gu1nness committed:
Merge branch '6/edge' into feat/DPE-6715-ldap-manager

Gu1nness committed:
feat: LDAP integration testing (mongodb)

Gu1nness committed:
test: ldap sharding integration

Gu1nness committed:
chore: move things around for easier integration

Gu1nness committed:
feat: add tasks

Gu1nness committed:
feat: mongos test

Gu1nness committed:
fix: try to fix all tests

Gu1nness committed:
fix: move kubernetes_model fixture + abort_on_fail

Gu1nness committed:
feat: update interface to provide managers for to reduce complexity

Gu1nness committed:
fix: create offer correctly

Gu1nness committed:
fix: offer fixes + mongos vm lxd

Gu1nness committed:
fix: try to fix PG storage issue

Gu1nness committed:
fix: series for data integrator

Gu1nness committed:
fix: small workflow fixes

Gu1nness committed:
fix: testing profile for postgres

Gu1nness committed:
fix: use new argument in create groups

Gu1nness committed:
fix: unit ids

Gu1nness committed:
fix: No direct connection

Gu1nness committed:
feat: connection string

Gu1nness committed:
fix: test collection name

Gu1nness committed:
fix: last fixes

Gu1nness committed:
fix: another round of fixes

Gu1nness committed:
fix: ldapQuery

Gu1nness committed:
fix: ldapQuery (bis)

Gu1nness committed:
fix: small typo

Gu1nness committed:
fix: connect to mongos

Gu1nness committed:
fix: bugfix userToDNMapping

Gu1nness committed:
fix: small changes

Gu1nness committed:
fix mongos config

Gu1nness committed:
fix postgres storage issue

Gu1nness committed:
fix: unit get on subordinate + permissions

Gu1nness committed:
fix: remove run on push on feat/**

Gu1nness committed:
fix: use subordinate + mongos container

Gu1nness committed:
fix: small fixes

Gu1nness committed:
fix: status message invalid parsing

Gu1nness committed:
fix: small changes

Gu1nness committed:
fix: remove hash on relation removed

Gu1nness committed:
fix: remove useless test

Gu1nness committed:
fix: operations order

Gu1nness committed:
feat: always restart on tls changes

Gu1nness committed:
fix: post review

Gu1nness committed:
fix: reconfigure workload on upgrade

Gu1nness committed:
fix: restart only if we're not waiting for new certs

Gu1nness committed:
fix: post-review

Mehdi-Bendriss committed:
Merge pull request #36 from canonical/feat/ldap-integration-testing

feat(DPE-6724): LDAP Integration testing

Gu1nness committed:
Merge branch '6/edge' into feat/DPE-6715-ldap-manager

Mehdi-Bendriss reviewed (approved):
Thank you Neha!
